src/reporter.cc

```diff
@@ -72,45 +90,93 @@ namespace {
+      std::string("multipart/mixed; boundary=") + kMultipartBoundary;
   http::client client;
   http::client::request request(endpoint);
-  std::string request_body = update_metadata_request->ToString();
   request << boost::network::header("User-Agent", user_agent);
-  request << boost::network::header("Content-Length",
-                                    std::to_string(request_body.size()));
-  request << boost::network::header("Content-Type", "application/json");
+  request << boost::network::header("Content-Type", content_type);
   request << boost::network::header("Authorization", auth_header);
-  request << boost::network::body(request_body);
+  request << boost::network::header("Expect", "100-continue");
+  std::ostringstream out;
+  out << std::endl;
+
+  for (json::value& entry : entries) {
+    const json::Object* single_request = entry->As<json::Object>();
+    const std::string request_body = single_request->ToString();
+    out << format::Substitute(kPartFormat, {
+      {"boundary", kMultipartBoundary},
+      {"name", single_request->Get<json::String>("name")},
+      {"path", publish_path},
+      {"length", std::to_string(request_body.size())},
+      {"body", request_body},
+    });
+  }
+  out << "--" << kMultipartBoundary << "--" << std::endl;
+  std::string multipart_body = out.str();
+  const int total_length = multipart_body.size();
+
+  request << boost::network::header("Content-Length",
+                                    std::to_string(total_length));
+  request << boost::network::body(multipart_body);
+
+  if (verbose_logging) {
+    LOG(INFO) << "About to send request: POST " << endpoint;
+    LOG(INFO) << "Headers: " << request.headers();
+    LOG(INFO) << "Body:" << std::endl << body(request);
+  }
+
   http::client::response response = client.post(request);
   if (status(response) >= 300) {
     throw boost::system::system_error(
-        boost::system::errc::make_error_code(boost::system::errc::not_connected),
+        boost::system::errc::make_error_code(
+            boost::system::errc::not_connected),
         format::Substitute("Server responded with '{{message}}' ({{code}})",
                            {{"message", status_message(response)},
                             {"code", format::str(status(response))}}));
   }
   if (verbose_logging) {
-    LOG(INFO) << "Server responded with " << body(response);
+    LOG(INFO) << format::Substitute(
+        "Server responded with '{{message}}' ({{code}})",
+        {{"message", status_message(response)},
+         {"code", format::str(status(response))}});
+    LOG(INFO) << "Headers: " << response.headers();
+    LOG(INFO) << "Body:" << std::endl << body(response);
   }
   // TODO: process response.
 }
 
 }
 
+std::string MetadataReporter::FullyQualifiedResourceLocation(
+    const std::string& location) const {
+  int num_dashes = std::count(location.begin(), location.end(), '-');
+  const std::string location_type = num_dashes == 2 ? "zones" : "regions";
```

Comment:
Just calling this out, the other usage of this ternary operator switches between zones and locations, while this one switches between zones and regions, is this expected? Ideally, it would be nice to have this check extracted into a helper function as something like "IsLocationZonal".

Reply:
This is because in the other place, we are trying to make the full name of the GKE cluster.
In GKE, locations/{region} is used in the regional cluster full names, and zones/{zone} is used in the zonal cluster full name.

Here, we are constructing the full location name based on the metadata API requirements.

Added a comment here as well.